src/InteractiveWindow/Editor/InteractiveWindow.cs

```diff
@@ -437,51 +403,49 @@ private void ClearView(bool insertInputPrompt)
 
         public void InsertCode(string text)
         {
-            if (!CheckAccess())
+            if (!OnUIThread())
             {
                 Dispatcher.BeginInvoke(new Action(() => InsertCode(text)));
                 return;
             }
 
+            AssertOnUIThread();
+
             if (_stdInputStart == null)
             {
-                if (_isRunning)
+                return;
+            }
+
+            if (_state == State.ExecutingInput)
+            {
+                AppendUncommittedInput(text);
+            }
+            else
+            {
+                if (!_textView.Selection.IsEmpty)
                 {
-                    AppendUncommittedInput(text);
+                    CutOrDeleteSelection(isCut: false);
                 }
-                else
-                {
-                    if (!TextView.Selection.IsEmpty)
-                    {
-                        CutOrDeleteSelection(isCut: false);
-                    }
 
-                    _editorOperations.InsertText(text);
-                }
+                _editorOperations.InsertText(text);
             }
         }
 
-        public Task SubmitAsync(IEnumerable<string> inputs)
+        Task IInteractiveWindow.SubmitAsync(IEnumerable<string> inputs)
         {
             var completion = new TaskCompletionSource<object>();
             var submissions = inputs.ToArray();
-            PendingSubmission[] pendingSubmissions = new PendingSubmission[submissions.Length];
-            if (submissions.Length == 0)
+            var numSubmissions = submissions.Length;
+            PendingSubmission[] pendingSubmissions = new PendingSubmission[numSubmissions];
```

Comment:
I suspect not, but Submit isn't (totally) a no-op when called with a zero-length array, so I'd prefer to refactor that separately.